aten/src/ATen/Error.h

```diff
@@ -0,0 +1,43 @@
+#pragma once
+
+#include <ATen/ATenGeneral.h> // for AT_API
+
+#include <cstdint>
+#include <cstdio>
+#include <stdexcept>
+#include <string>
+
+namespace at {
+namespace detail {
+struct SourceLocation {
+  const char* function;
+  const char* file;
+  uint32_t line;
+};
+} // namespace detail
+
+template <typename... FormatArgs>
+[[noreturn]] AT_API void error(
```

Comment:
This will create a symbol per every different type configuration when calling `error`, but I guess that's ok 😕

Reply:
What happens if I pass a `std::string` to this argument? It seems weird to templatize, but then for C++ types to not work. (This, btw, is why I have been somewhat reluctant to keep adding printf-style error sites; it's a loaded footgun.)

Reply:
We could use the stringstream-like errors, just like we do in `ErrorReport` in the script. They should be safer.

Reply:
More random header-only libraries that may make this easier / more readable: https://github.com/fmtlib/fmt

Reply:
I don't think that adds a lot of value on top of the `<<` API to justify adding yet another library to dependencies.

Reply:
So, are we ok with losing the `printf`-like behavior? Then we wouldn't need a macro. We'd just need to rewrite e.g. `AT_ERROR("%s%sType is not enabled.",toString(p),toString(s));` to `throw at::error() << toString(p) <<  toString(s) << "Type is not enabled.";`.

As for the concern about `std::string`, we could add something like `enable_if<is_literal_type<FormatArgs>>`

Reply:
Actually the current error message when you pass a non-trivial type is pretty decent: 
```
/home/psag/pytorch/pytorch/aten/src/ATen/Error.h:19:65: error: cannot pass objects of non-trivially-copyable type ‘class std::basic_string<char>’ through ‘...’
   snprintf(buffer, sizeof(buffer), format_string, format_args...);
```